File: NationalCode.php
<?php

/**
 * Validates an Iranian National Code based on the official algorithm.
 *
 * This function checks the validity of an Iranian National Code by:
 * 1. Ensuring the input is 8 to 10 digits long and numeric
 * 2. Standardizing it to 10 digits by padding with leading zeros
 * 3. Checking for repeating digits (e.g., 1111111111 is invalid)
 * 4. Calculating the check digit using the official algorithm:
 *    - Multiply each of the first 9 digits by weights 10 down to 2
 *    - Sum the results and compute remainder when divided by 11
 *    - If remainder < 2, it’s the check digit; otherwise, 11 - remainder
 * 5. Comparing the calculated check digit with the provided one
 *
 * @param string $nationalCode The national code to validate (can include non-numeric characters)
 * @return bool Returns true if the code is valid, false otherwise
 */
function validateIranianNationalCode(string $nationalCode = ''): bool
{
    // Remove non-numeric characters from the input
    $cleanedCode = preg_replace('/[^0-9]/', '', (string) $nationalCode);
    $codeLength = strlen($cleanedCode);

    // Check if length is between 8 and 10 digits
    if ($codeLength < 8 || $codeLength > 10) {
        return false;
    }

    // Standardize to 10 digits by padding with leading zeros
    $standardizedCode = str_pad($cleanedCode, 10, '0', STR_PAD_LEFT);
    $digits = str_split($standardizedCode);

    // Check if all digits are the same (invalid case)
    if (count(array_unique($digits)) === 1) {
        return false;
    }

    // Extract the check digit (last digit)
    $checkDigit = (int) array_pop($digits);

    // Calculate weighted sum (weights: 10 down to 2)
    $sum = 0;
    for ($i = 0; $i < 9; $i++) {
        $sum += (int) $digits[$i] * (10 - $i);
    }

    // Compute remainder and check digit
    $remainder = $sum % 11;
    $calculatedCheckDigit = $remainder < 2 ? $remainder : 11 - $remainder;

    // Return true if calculated check digit matches the provided one
    return $calculatedCheckDigit === $checkDigit;
}

// Test cases
var_dump(validateIranianNationalCode('1111111111')); // false (repeating digits)
var_dump(validateIranianNationalCode('12345'));      // false (too short)

File: NationalCode.py
import re

def validate_iranian_national_code(national_code: str) -> bool:
    # Remove non-numeric characters
    cleaned_code = re.sub(r'[^0-9]', '', national_code)
    code_length = len(cleaned_code)

    # Check if length is between 8 and 10 digits
    if code_length < 8 or code_length > 10:
        return False

    # Standardize to 10 digits by padding with leading zeros
    standardized_code = cleaned_code.zfill(10)
    digits = list(map(int, standardized_code))

    # Check if all digits are the same (invalid case)
    if len(set(digits)) == 1:
        return False

    # Extract the check digit (last digit)
    check_digit = digits.pop()

    # Calculate weighted sum (weights: 10 down to 2)
    weighted_sum = sum(digits[i] * (10 - i) for i in range(9))

    # Compute remainder and check digit
    remainder = weighted_sum % 11
    calculated_check_digit = remainder if remainder < 2 else 11 - remainder

    # Return true if calculated check digit matches the provided one
    return calculated_check_digit == check_digit

# Test cases
print(validate_iranian_national_code('1111111111'))  # False (repeating digits)
print(validate_iranian_national_code('12345'))       # False (too short)
